File: 2024/day6/6.py
from collections import defaultdict


def getInput():
    f = open("input", "r")

    board = []

    for line in f:
        board.append(list(line[: len(line) - 1]))

    f.close()
    return board


def getPos(input):
    for r in range(len(input)):
        for c in range(len(input[0])):
            if input[r][c] == "^":
                return [r, c]

    return [-1, -1]


def getMoveSum(input):
    sum = 0
    for r in range(len(input)):
        for c in range(len(input[0])):
            if input[r][c] == "X":
                sum += 1

    return sum


def part1(input):
    pos = getPos(input)
    direction = 0
    m, n = len(input), len(input[0])

    while pos[0] >= 0 and pos[0] < m and pos[1] >= 0 and pos[1] < n:
        input[pos[0]][pos[1]] = "X"

        match direction:
            case 0:
                if pos[0] - 1 >= 0 and input[pos[0] - 1][pos[1]] == "#":
                    direction = 1
                else:
                    pos[0] -= 1
            case 1:
                if pos[1] + 1 < n and input[pos[0]][pos[1] + 1] == "#":
                    direction = 2
                else:
                    pos[1] += 1
            case 2:
                if pos[0] + 1 < m and input[pos[0] + 1][pos[1]] == "#":
                    direction = 3
                else:
                    pos[0] += 1
            case 3:
                if pos[1] - 1 >= 0 and input[pos[0]][pos[1] - 1] == "#":
                    direction = 0
                else:
                    pos[1] -= 1

    return getMoveSum(input)


def part2(input):
    m, n = len(input), len(input[0])
    startPos = getPos(input)

    sum = 0
    for r in range(m):
        for c in range(n):
            if input[r][c] == "#" or (r == startPos[0] and c == startPos[1]):
                continue

            input[r][c] = "O"
            count = 0
            direction = 0
            pos = [startPos[0], startPos[1]]

            while (
                pos[0] >= 0
                and pos[0] < m
                and pos[1] >= 0
                and pos[1] < n
                and count < m * n
            ):
                count += 1

                match direction:
                    case 0:
                        if pos[0] - 1 >= 0 and input[pos[0] - 1][pos[1]] in "#O":
                            direction = 1
                        else:
                            pos[0] -= 1
                    case 1:
                        if pos[1] + 1 < n and input[pos[0]][pos[1] + 1] in "#O":
                            direction = 2
                        else:
                            pos[1] += 1
                    case 2:
                        if pos[0] + 1 < m and input[pos[0] + 1][pos[1]] in "#O":
                            direction = 3
                        else:
                            pos[0] += 1
                    case 3:
                        if pos[1] - 1 >= 0 and input[pos[0]][pos[1] - 1] in "#O":
                            direction = 0
                        else:
                            pos[1] -= 1

            print(count)
            if count >= m * n:
                sum += 1

            input[r][c] = "."

    return sum


def main():
    board = getInput()

    # print(part1(board))
    print(part2(board))


main()

File: 2024/day6/6.cpp
#include <bits/stdc++.h>
#include <cmath>
#include <fstream>
#include <iostream>
#include <unordered_map>

std::vector<std::string> openInput(char *filename) {
  std::ifstream f(filename);

  if (!f.is_open()) {
    std::cout << "Cant open file\n";
    std::abort();
  }

  std::string line;
  std::vector<std::string> input;

  while (std::getline(f, line)) {
    input.push_back(line);
  }

  return input;
};

std::pair<int, int> getGuardPos(std::vector<std::string> input) {
  for (int r = 0; r < input.size(); r++) {
    for (int c = 0; c < input[0].size(); c++) {
      if (input[r][c] == '^') {
        return {r, c};
      }
    }
  }
  return {-1, -1};
}

int getUniqueSpots(std::vector<std::string> input) {
  int sum = 0;

  for (int r = 0; r < input.size(); r++) {
    for (int c = 0; c < input[0].size(); c++) {
      if (input[r][c] == 'X') {
        sum++;
      }
    }
  }

  return sum;
}

int part1(std::vector<std::string> input) {
  std::pair<int, int> pos = getGuardPos(input);
  int m = input.size();
  int n = input[0].size();

  // direct 0 = up 1 = right 2 = down 3 = left
  int direction = 0;
  while (pos.first >= 0 && pos.first < m && pos.second >= 0 && pos.second < n) {

    input[pos.first][pos.second] = 'X';
    switch (direction) {
    case 0:
      if (pos.first - 1 >= 0 && input[pos.first - 1][pos.second] == '#') {
        direction = 1;
      } else {
        pos.first--;
      }
      break;
    case 1:
      if (pos.second + 1 < n && input[pos.first][pos.second + 1] == '#') {
        direction = 2;
      } else {
        pos.second++;
      }
      break;
    case 2:
      if (pos.first + 1 < m && input[pos.first + 1][pos.second] == '#') {
        direction = 3;
      } else {
        pos.first++;
      }
      break;
    case 3:
      if (pos.second - 1 >= 0 && input[pos.first][pos.second - 1] == '#') {
        direction = 0;
      } else {
        pos.second--;
      }
      break;
    }
  };

  return getUniqueSpots(input);
}

int part2(std::vector<std::string> input) {
  std::pair<int, int> startPos = getGuardPos(input);

  int sum = 0;
  int m = input.size();
  int n = input[0].size();

  for (int r = 0; r < m; r++) {
    for (int c = 0; c < n; c++) {
      if (input[r][c] == '#' || (r == startPos.first && c == startPos.second)) {
        continue;
      }

      input[r][c] = 'O';
      std::pair<int, int> pos = {startPos.first, startPos.second};
      int direction = 0;
      int count = 0;

      while (pos.first >= 0 && pos.first < m && pos.second >= 0 &&
             pos.second < n && count < m * n) {
        input[pos.first][pos.second] = 'X';

        count++;

        switch (direction) {
        case 0:
          if (pos.first - 1 >= 0 && (input[pos.first - 1][pos.second] == '#' ||
                                     input[pos.first - 1][pos.second] == 'O')) {
            direction = 1;
          } else {
            pos.first--;
          }
          break;
        case 1:
          if (pos.second + 1 < n && (input[pos.first][pos.second + 1] == '#' ||
                                     input[pos.first][pos.second + 1] == 'O')) {
            direction = 2;
          } else {
            pos.second++;
          }
          break;
        case 2:
          if (pos.first + 1 < m && (input[pos.first + 1][pos.second] == '#' ||
                                    input[pos.first + 1][pos.second] == 'O')) {
            direction = 3;
          } else {
            pos.first++;
          }
          break;
        case 3:
          if (pos.second - 1 >= 0 && input[pos.first][pos.second - 1] == '#' ||
              (input[pos.first][pos.second - 1] == 'O')) {
            direction = 0;
          } else {
            pos.second--;
          }
          break;
        }
      }

      if (count >= m * n) {
        sum++;
      }

      input[r][c] = '.';
    }
  }

  return sum;
}

int main(int argc, char *argv[]) {
  if (argc != 2) {
    std::cout << "Must add filename\n";
    return 0;
  }

  std::vector<std::string> input;
  input = openInput(argv[1]);

  // std::cout << part1(input) << "\n";
  std::cout << part2(input) << "\n";

  return 0;
}
